### ISSUE TITLE
packer hcl2_upgrade crashing

### ISSUE THREAD

**Issue by mjbright** [hashicorp/packer#12199]
#### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

#### Overview of the Issue

I'm trying to create a ubuntu2204 template for Proxmox using the config at 
https://github.com/thedatabaseme/cookbooks/tree/master/packer/proxmox

But Packer crashes indicating a likely bug.

( I also experienced a similar crash using https://github.com/Aaron-K-T-Berry/packer-ubuntu-proxmox-template )

```
packer hcl2_upgrade ubuntu2204.json
panic: assignment to entry in nil map

goroutine 1 [running]:
github.com/hashicorp/packer/command.(*VariableParser).Parse(0xc00091f700, 0xc0001433f0)
        /home/runner/work/packer/packer/command/hcl2_upgrade.go:881 +0x3d2
github.com/hashicorp/packer/command.(*HCL2UpgradeCommand).RunContext(0xc000ba47e0, {0xc000120170?, 0x1?}, 0xc000198090)
        /home/runner/work/packer/packer/command/hcl2_upgrade.go:181 +0x494
github.com/hashicorp/packer/command.(*HCL2UpgradeCommand).Run(0xc000ba47e0, {0xc000120170, 0x1, 0x1})
        /home/runner/work/packer/packer/command/hcl2_upgrade.go:42 +0xc5
github.com/mitchellh/cli.(*CLI).Run(0xc000a74780)
        /home/runner/go/pkg/mod/github.com/mitchellh/cli@v1.1.2/cli.go:262 +0x5f8
main.wrappedMain()
        /home/runner/work/packer/packer/main.go:262 +0xb78
main.realMain()
        /home/runner/work/packer/packer/main.go:49 +0xf3
main.main()
        /home/runner/work/packer/packer/main.go:35 +0x19



!!!!!!!!!!!!!!!!!!!!!!!!!!! PACKER CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Packer crashed! This is always indicative of a bug within Packer.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Packer[1] so that we can fix this.

[1]: https://github.com/hashicorp/packer/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! PACKER CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

#### Reproduction Steps

Check out files from https://github.com/thedatabaseme/cookbooks/tree/master/packer/proxmox

Try to upgrade the main config file to hcl2 as ```packer hcl2_upgrade ubuntu2204.json```

### Packer version

```Packer v1.8.5```

### Simplified Packer Template

Not sure what simplifications would cause other breakage

ubuntu2204.json is only 47 lines without complex structure

### Operating system and Environment details

Running on Ubuntu 20.04 LTS

### Log Fragments and crash.log files

crash.log:
```
2023/01/17 09:24:23 [INFO] Packer version: 1.8.5 [go1.18.9 linux amd64]
2023/01/17 09:24:23 [TRACE] discovering plugins in /home/mjb/usr/bin
2023/01/17 09:24:23 [TRACE] discovering plugins in /home/mjb/.config/packer/plugins
2023/01/17 09:24:23 [TRACE] discovering plugins in .
2023/01/17 09:24:23 [INFO] PACKER_CONFIG env var not set; checking the default config file path
2023/01/17 09:24:23 [INFO] PACKER_CONFIG env var set; attempting to open config file: /home/mjb/.packerconfig
2023/01/17 09:24:23 [WARN] Config file doesn't exist: /home/mjb/.packerconfig
2023/01/17 09:24:23 [INFO] Setting cache directory: /home/mjb/.cache/packer
2023/01/17 09:24:23 waiting for all plugin processes to complete...
panic: assignment to entry in nil map

goroutine 1 [running]:
github.com/hashicorp/packer/command.(*VariableParser).Parse(0xc000c73b00, 0xc00044ec60)
        /home/runner/work/packer/packer/command/hcl2_upgrade.go:881 +0x3d2
github.com/hashicorp/packer/command.(*HCL2UpgradeCommand).RunContext(0xc00076ca50, {0xc00006e1a0?, 0x1?}, 0xc00053e120)
        /home/runner/work/packer/packer/command/hcl2_upgrade.go:181 +0x494
github.com/hashicorp/packer/command.(*HCL2UpgradeCommand).Run(0xc00076ca50, {0xc00006e1a0, 0x1, 0x1})
        /home/runner/work/packer/packer/command/hcl2_upgrade.go:42 +0xc5
github.com/mitchellh/cli.(*CLI).Run(0xc0005f3cc0)
        /home/runner/go/pkg/mod/github.com/mitchellh/cli@v1.1.2/cli.go:262 +0x5f8
main.wrappedMain()
        /home/runner/work/packer/packer/main.go:262 +0xb78
main.realMain()
        /home/runner/work/packer/packer/main.go:49 +0xf3
main.main()
        /home/runner/work/packer/packer/main.go:35 +0x19
```

**Comment by nywilken** [comment]
@mjbright thanks for reporting this issue. It looks like the new logic for auto creating variable blocks for referenced variables coming from an external variables file is failing because there is no existing variables block in the source JSON template. I'll work on getting this resolved shortly. 


Relevant information from crash log. 

```
panic: assignment to entry in nil map

goroutine 1 [running]:
github.com/hashicorp/packer/command.(*VariableParser).Parse(0xc000c73b00, 0xc00044ec60)
        /home/runner/work/packer/packer/command/hcl2_upgrade.go:881 +0x3d2
```

**Comment by nywilken** [comment]
Hi @mjbright the fix for this was merged and will be in the next 1.8.6 release. If you are currently stuck on upgrading the template you have a few options until the fix is released. 

The issue stems from the fact that there is no variables stanza in the source JSON template. You can add the variables stanza to the top of the ubuntu2204.json JSON template to prevent the command from crashing. Having this stanza in the template will ensure that the variables map is not nil, which is causing the crash. 

```json
"variables": {
   "unsused": null
},
...
```

Alternatively you can declare all of the variables in ubuntu2204.json JSON template, which is what will happen when you convert the the template to HCL2. 

```json
"variables": {
  "proxmox_hostname": null,
  "proxmox_username": null,
  "proxmox_password": null,
  "proxmox_node":     null,
  "version_tag":      null,
  "clone_vm":         null,
  "ssh_username":    null,
  "ssh_password":     null
}
```

Or you can try the latest nightly once released later today. The fix will be in the nightly https://github.com/hashicorp/packer/releases/tag/nightly


Example of what the generated HCL will look like `packer hcl2_upgrade --with-annotations ubuntu2204.json`

```hcl
# This file was autogenerated by the 'packer hcl2_upgrade' command. We
# recommend double checking that everything is correct before going forward. We
# also recommend treating this file as disposable. The HCL2 blocks in this
# file can be moved to other files. For example, the variable blocks could be
# moved to their own 'variables.pkr.hcl' file, etc. Those files need to be
# suffixed with '.pkr.hcl' to be visible to Packer. To use multiple files at
# once they also need to be in the same folder. 'packer inspect folder/'
# will describe to you what is in that folder.

# Avoid mixing go templating calls ( for example ```{{ upper(`string`) }}``` )
# and HCL2 calls (for example '${ var.string_value_example }' ). They won't be
# executed together and the outcome will be unknown.

# All generated input variables will be of 'string' type as this is how Packer JSON
# views them; you can change their type later on. Read the variables type
# constraints documentation
# https://www.packer.io/docs/templates/hcl_templates/variables#type-constraints for more info.
variable "clone_vm" {
  type = string
}

variable "proxmox_hostname" {
  type = string
}

variable "proxmox_node" {
  type = string
}

variable "proxmox_password" {
  type = string
}

variable "proxmox_username" {
  type = string
}

variable "ssh_password" {
  type = string
}

variable "ssh_username" {
  type = string
}

variable "version_tag" {
  type = string
}

# source blocks are generated from your builders; a source can be referenced in
# build blocks. A build block runs provisioner and post-processors on a
# source. Read the documentation for source blocks here:
# https://www.packer.io/docs/templates/hcl_templates/blocks/source
source "proxmox-clone" "autogenerated_1" {
  clone_vm                 = "${var.clone_vm}"
  cores                    = 1
  full_clone               = false
  insecure_skip_tls_verify = true
  memory                   = 2048
  network_adapters {
    bridge = "vmbr0"
  }
  node                 = "${var.proxmox_node}"
  os                   = "l26"
  password             = "${var.proxmox_password}"
  proxmox_url          = "https://${var.proxmox_hostname}/api2/json"
  sockets              = 1
  ssh_password         = "${var.ssh_password}"
  ssh_timeout          = "90m"
  ssh_username         = "${var.ssh_username}"
  task_timeout         = "10m"
  template_description = "image made from packer"
  template_name        = "${var.clone_vm}${var.version_tag}"
  username             = "${var.proxmox_username}"
}

# a build block invokes sources and runs provisioning steps on them. The
# documentation for build blocks can be found here:
# https://www.packer.io/docs/templates/hcl_templates/blocks/build
build {
  sources = ["source.proxmox-clone.autogenerated_1"]

  provisioner "shell" {
    inline       = ["echo ${var.version_tag} > provision.txt", "sudo apt update", "sudo apt -y upgrade", "sudo apt -y install apache2", "sudo apt -y install qemu-guest-agent cloud-init"]
    pause_before = "20s"
  }

}
```
